### ISSUE TITLE
[BUG] Helm: Default settings do not work

### ISSUE THREAD

**Issue by klauserber** [longhorn/longhorn#8145]
## Describe the bug

On a fresh Longhorn install with helm the default settings do not take effect

## To Reproduce

I haved installed longhorn with helm and these values:

```
defaultSettings:
  allowRecurringJobWhileVolumeDetached: enabled
  backupTarget: s3://backup-longhorn-bm-isium-de/
  backupTargetCredentialSecret: longhorn-backup-secret
  recurringJobMaxRetention: 14
  storageMinimalAvailablePercentage: 5
ingress:
  enabled: true
  host: longhorn.bm.dc1.isium.de
  ingressClassName: nginx-internal
persistence:
  defaultDataLocality: best-effort
```

## Expected behavior

The settings should be appear in the Longhorn UI
 
## Support bundle for troubleshooting

[supportbundle_4bdc31e4-aaf6-4450-add0-84e8f7054857_2024-03-09T14-39-19Z.zip](https://github.com/longhorn/longhorn/files/14546644/supportbundle_4bdc31e4-aaf6-4450-add0-84e8f7054857_2024-03-09T14-39-19Z.zip)

## Environment

 - Longhorn version: 1.6.0
 - Impacted volume (PV): none
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE2
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 3
 - Node config
   - OS type and version: Ubuntu
   - Kernel version: 5.15.0-100-generic
   - CPU per node: 8
   - Memory per node: 96 GB
   - Disk type (e.g. SSD/NVMe/HDD): SSD
   - Network bandwidth between the nodes (Gbps): 10
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Baremetal
 - Number of Longhorn volumes in the cluster: 0

## Additional context

The Configmap `longhorn-default-setting` was correctly created. But when I dig into the templates of the Chart, there is no other resource, that reference the Configmap. How do this work? Is there anything missing?

**Comment by klauserber** [comment]
[supportbundle_4bdc31e4-aaf6-4450-add0-84e8f7054857_2024-03-09T14-39-19Z.zip](https://github.com/longhorn/longhorn/files/14546641/supportbundle_4bdc31e4-aaf6-4450-add0-84e8f7054857_2024-03-09T14-39-19Z.zip)

**Comment by ejweber** [comment]
There is a controller within longhorn-manager that syncs the settings from the ConfigMap to Settings CRs such that they can be seen from the UI, etc.

In your case, it looks like this controller is rejecting one of the configured values. [allowRecurringJobWhileVolumeDetached / allow-recurring-job-while-volume-detached](https://longhorn.io/docs/1.6.0/references/settings/#allow-recurring-job-while-volume-is-detached) should be `true` or `false`. 

```
2024-03-09T13:41:34.526352252Z time="2024-03-09T13:41:34Z" level=error msg="Invalid value enabled for the boolean setting allow-recurring-job-while-volume-detached" func=types.GetCustomizedDefaultSettings file="setting.go:1587" error="strconv.ParseBool: parsing \"enabled\": invalid syntax"
```

Somewhat surprisingly, if one ConfigMap setting is incorrect, Longhorn will refuse to reconcile all ConfigMap settings. TBH, I'm not sure why this is the case off the top of my head, but it looks quite intentional.

https://github.com/longhorn/longhorn-manager/blob/edf23eddc3b6c307031eaa770c3d312c963f25a5/types/setting.go#L1584-L1592

So I think correcting the incorrect setting will get things working for you. Please let us know how it goes.

**Comment by klauserber** [comment]
I have changed the setting and it works like a charme. Thank you very much.

Some kind of validation would be helpful.

**Comment by ejweber** [comment]
Thanks for the update @klauserber. I was writing up a feature suggestion in response to your feedback that would use the webhook built into longhorn-manager to validate the `longhorn-default-settings` ConfigMap using the existing validation code around https://github.com/longhorn/longhorn-manager/blob/edf23eddc3b6c307031eaa770c3d312c963f25a5/types/setting.go#L1584-L1592. The idea was that the webhook should reject a ConfigMap create or update if it includes invalid settings. The problem is that the `longhorn-default-setting` ConfigMap:

- Is most often created BEFORE and longhorn-manager are running, when it cannot be validated by the webhook.
- Is most often modified in the BEGINNING of an upgrade, when it will be validated by the old longhorn-manager (which may not know about new settings).

I considered adding a JSON schema to the Helm chart, but:

- This would require us to maintain two sets of validation logic: the JSON schema in the chart and the Go logic that already exists.
- kubectl deployments would not benefit from it.

I considered adding the Go validation to the Helm pre-upgrade job (and adding an additional pre-install job that doesn't currently exist), but:

- kubectl deployments would not benefit from it.

In short, while I recognize that the root cause of the issue was difficult to determine in this situation, I don't see much of a good path towards improving it right now.

I'm going to close this one out, since it has been resolved. If you disagree with my analysis or feel strongly that we need to make an improvement in this area, please feel free to open a feature request anyway and link it back to this issue. We will prioritize it based on the traction it gets.

**Comment by klauserber** [comment]
It is absoluty ok to close the issue, because it is fixed.

I think I have overlooked it in the controller log, probably there was a error about this wrong Value.
